### ISSUE TITLE
Not working properly with CUDA

### ISSUE THREAD

**Issue by fedelopez77** [geoopt/geoopt#81]
I am working with a model similar to the example below:

```python
class Model(nn.Module):
    def __init__(self, word2vec):
        super(Model, self).__init__()
        self.word_lut = gt.ManifoldParameter(word2vec, manifold=gt.PoincareBall())
``` 

When I move the model to CUDA, and run it there is a problem during the optimization (See full traceback below [1]):
`RuntimeError: Expected object of backend CUDA but got backend CPU for argument #2 'other'`

The reason of the error is that when I move the model to CUDA, the tensor in the ManifoldParameter is moved, but its manifold is not, therefore the attribute tensor `self.c` of the manifold is still allocated in the CPU.

I solved it doing this:
```python
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")    
for p in model.parameters():
    if isinstance(p, (ManifoldParameter, ManifoldTensor)):
        p.manifold.to(device)
```

I don't know if issue #49 was an attempt to solve this and I didn't moved the model in the proper way to CUDA or this is not completely documented. In case that the explanation on how to do this properly exists and I couldn't find it, my apologies for creating this issue.

[1] Full traceback
```
Traceback (most recent call last):
  File "./train.py", line 122, in <module>
    main()
  File "./train.py", line 113, in main
    coach.train()
  File "/home/lopezfo/projects/hyfi/hyfi/Coach.py", line 43, in train
    train_loss = self.train_epoch(epoch)
  File "/home/lopezfo/projects/hyfi/hyfi/Coach.py", line 99, in train_epoch
    self.optim.step()
  File "/home/lopezfo/anaconda3/envs/hyfi/lib/python3.6/site-packages/geoopt/optim/radam.py", line 145, in step
    state["exp_avg_sq"],
  File "/home/lopezfo/anaconda3/envs/hyfi/lib/python3.6/site-packages/geoopt/optim/tracing.py", line 34, in partial
    step(manifold, *args, **kwargs)
  File "/home/lopezfo/anaconda3/envs/hyfi/lib/python3.6/site-packages/geoopt/optim/radam.py", line 191, in perform_step
    point, -step_size * direction, exp_avg
  File "/home/lopezfo/anaconda3/envs/hyfi/lib/python3.6/site-packages/geoopt/manifolds/poincare/__init__.py", line 110, in retr_transp
    y = self.retr(x, u)
  File "/home/lopezfo/anaconda3/envs/hyfi/lib/python3.6/site-packages/geoopt/manifolds/poincare/__init__.py", line 69, in retr
    return math.project(approx, c=self.c)
  File "/home/lopezfo/anaconda3/envs/hyfi/lib/python3.6/site-packages/geoopt/manifolds/poincare/math.py", line 78, in project
    return _project(x, c, dim, eps)
  File "/home/lopezfo/anaconda3/envs/hyfi/lib/python3.6/site-packages/geoopt/manifolds/poincare/math.py", line 86, in _project
    cond = norm > maxnorm
RuntimeError: Expected object of backend CUDA but got backend CPU for argument #2 'other'
```

**Comment by rrkarim** [comment]
It is because of recent updates, has an easy fix.

**Comment by rrkarim** [comment]
I've checked the `Poincare ball` methods and it seems like the bug affects only the projection.

**Comment by ferrine** [comment]
The proper solution is to additionally attach the ball to the module attribute. This will make the conversion work. So far this is the only way to do it

```
self.ball = ball
```

**Comment by rrkarim** [comment]
Undocumented behavior, we need to handle the exception somewhere.

**Comment by ferrine** [comment]
True

**Comment by ferrine** [comment]
What about documenting things instead? I do not like the idea catching exceptions in all downstream methods

**Comment by rrkarim** [comment]
I have no idea how we can document it other than giving the example of how to construct the model on a ball.

**Comment by ferrine** [comment]
@fedelopez77, did the advice solve your issue?

**Comment by fedelopez77** [comment]
@ferrine I am following [this code](https://github.com/bcol23/HyperIM/blob/master/net/HyperIM.py#L20) as an example, and they have "many" Poincaré balls. 
Is that incorrect? Should I only create one instance of the ball and pass it around for all the ManifoldParameters that I want to optimize in the ball?

**Comment by ferrine** [comment]
@fedelopez77 This code does not seem to work on gpu. There are some problems with pytorch integration and our design. For instance, a manifold can have its own parameters (like sphere on a subspace or Poincare ball) and we need to track their device. That was the main reason for making them `torch.nn.Module`'s: `manifold.cuda()` seemed too appealing. However, we also need to attach a menifold to a `Parameter`/`Tensor`, to make optimizer aware about "how to make an update". Thus a manifold is an attribute of `{Manifold|Parameter}Tensor`. 

BUT, when we call `module.cuda()` it traverses all modules and their submodules looking for tensors and parameters, but nut modules of parameters(!). That is the core issue why a `Manifold`, attached to a `Tensor` is not visible for `.cuda()` method. The easiest solution is to attach `Manifold` to the parent `Module` to meke `.cuda()` work as expected.

Geoopt side solution is either to provide `geoopt.Module` with the correct implementation of `.cuda()`, etc, or patch `torch.nn.Module` directly. We did not solve this design problem as neither solution is a silver bullet and has disadvantages.

This is an [example](https://gist.github.com/ferrine/bdab3b2d34e1494421ac262b4be499df) of how we decided to work with modules, that have parameters on a manifold (PoincareBall). Having many "balls" might be error-prone, so we pass either `c` or shared `ball` instance to the module. In functions, we use methods of manifolds to be safe from typos.
